### ISSUE TITLE
fix(types): replace SVG <svg-path> placeholder with real path data parser

### ISSUE THREAD

**Issue by YusukeHirao** [markuplint/markuplint#3619]
## Summary

Add validation for the `d` attribute on SVG `<path>` — SVG path data syntax.

## Spec reference

- https://svgwg.org/svg2-draft/paths.html#TheDProperty

## Impact

~2 missed errors in the nu-html-checker compatibility benchmark.

## Benchmark cross-reference

_Sync'd from `tests/external/snapshots/` (submodule=1429313, nu-validator@sha256:59b0e97e2664, markuplint=5.0.0-rc.4, generated=2026-04-22 23:43:45 UTC)._

Filter: `html-svg\/(paths-data|masking-path|imp-path)` → **34** fixtures. Verdict tally: match-error=2, match-clean=0, ml-only=32, nu-only=0, nu-over=0.

Direction is inverted. The issue asks for "missing validation", but 32 of the 34 `html-svg/paths-data` / `masking-path` fixtures are `ml-only` — markuplint currently rejects SVG path data that nu-validator (and the SVG spec) accept. The fix is to tighten the `<svg-path>` type into a real parser that stops emitting false positives, not to add more strict checks. Tracking the real work in a fresh issue; this one is closed as "not planned".

| fixture | verdict | nu id / ml rule |
| --- | --- | --- |
| `html-svg/imp-path-01-f-isvalid.html` | ml-only | `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-01-b-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-02-b-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-03-b-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-04-b-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-05-f-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-06-b-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-07-b-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-08-b-isvalid.html` | ml-only | `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-09-b-isvalid.html` | ml-only | `@markuplint/ml-core` |
| `html-svg/masking-path-10-b-isvalid.html` | ml-only | `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-11-b-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-12-f-isvalid.html` | ml-only | `@markuplint/ml-core` |
| `html-svg/masking-path-13-f-isvalid.html` | ml-only | `invalid-attr`, `permitted-contents` |
| `html-svg/masking-path-14-f-isvalid.html` | ml-only | `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-01-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-02-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-03-f-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-04-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-05-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-06-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-07-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-08-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-09-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-10-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-12-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-13-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-14-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-15-t-isvalid.html` | ml-only | `deprecated-attr`, `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-16-t-isvalid.html` | ml-only | `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-17-f-isvalid.html` | ml-only | `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-18-f-novalid.html` | match-error | — |
| `html-svg/paths-data-19-f-isvalid.html` | ml-only | `invalid-attr`, `permitted-contents` |
| `html-svg/paths-data-20-f-novalid.html` | match-error | — |

Reproduce: see [`tests/external/CLAUDE.md`](https://github.com/markuplint/markuplint/blob/dev/tests/external/CLAUDE.md) "Auditing a claim against the benchmark".

**Comment by YusukeHirao** [comment]
## nu-html-checker 実出力での検証

### テスト
- `<svg><path d="M 10 10 L invalid"/></svg>` → nu: 1 error / ml: 0 error
- `<svg><path d="M 10 10 L 20 20"/></svg>` → nu: 0 / ml: 0

### 判定: CONFIRMED

nu-html-checker は SVG path data の完全な tokenizer parse を行うが、markuplint は文字列として受理する。`@markuplint/types` に SVG path data validator（新規）を実装する必要あり。

### 影響規模
Issue 記載の「~2 missed errors」は妥当。小〜中規模の実装。

**Comment by YusukeHirao** [comment]
## ⚠️ 過去コメントは未検証

本 Issue の 2026-04-14 以前のコメントは nu-html-checker を実行せずに書かれた推測ベースです。以下が Docker 実測結果です。

## 実測結果（2026-04-15）

### 全 6 ケース対比（`<path d="...">`）

| # | `d=` 値 | nu-html-checker | markuplint | Gap |
|---|---------|----------------|-----------|-----|
| 1 | `M 10 10 L invalid` | `Bad value "M 10 10 L invalid" for attribute "d" on element "path": Expected coordinate pair for "L" command but found "i" instead (context: "M 10 10 L i").` | （検出なし） | **YES** |
| 2 | `""` (空) | 違反なし | 「属性「d」はCSS構文として「<svg-path>」である必要があります」(`invalid-attr`) | **INVERTED**（markuplint 過剰） |
| 3 | `M 1 1 Z` (正常) | 違反なし | 違反なし | NO |
| 4 | `garbage` | `Bad value "garbage" ... Expected command but found "g" (context: "g").` | （検出なし） | **YES** |
| 5 | `M 10 10 C 20 20` (引数不足) | `Bad value ... Expected x2 coordinate for "C" command but value ended (context: "M 10 10 C 20 20").` | （検出なし） | **YES** |
| 6 | `M 10 10 L 20 20` (正常) | 違反なし | 違反なし | NO |

## 真のスコープ

markuplint の `<svg-path>` 型検証は**両方向に壊れている**:

- **過剰**: 空文字列 `d=""` を reject（SVG2 仕様では valid）
- **不足**: 任意の非空文字列を accept（真の path BNF パーサーなし）

## 実装方針

`@markuplint/types` の `<svg-path>` 型を SVG2 BNF（https://svgwg.org/svg2-draft/paths.html#PathDataBNF）に沿ったパーサーで再実装:

- 空文字列は valid（no path）として受け入れ
- 各コマンド（`M m L l H h V v C c S s Q q T t A a Z z`）を認識
- コマンドごとの引数 arity/形式を検証
- 数値形式を BNF どおりに検証

推奨メッセージ形式（nu に倣う）: `Bad value "<value>" for attribute "d" on element "path": <reason> (context: "<excerpt>").`

**Comment by YusukeHirao** [comment]
## 補足: 原因の詳細

前コメントで「両方向に壊れている」とだけ書いたが、コード確認で原因が判明。

### コード: `packages/@markuplint/types/src/css-defs.ts:176-185`

```typescript
'<svg-path>': {
  ref: 'https://svgwg.org/svg2-draft/paths.html#PathDataBNF',
  syntax: {
    apply: '<svg-path>',
    // Validation not yet implemented; accepts any value.
    def: {
      'svg-path': '<any-value>',
    },
  },
},
```

### 両方向のバグの原因

**過剰検出（`d=""` reject）**: CSS の `<any-value>` は「one or more preserved tokens」の定義で、空文字列を reject する。そのため `d=""` は SVG2 仕様上 valid にもかかわらず markuplint が拒否する。

**検出不足（garbage accept）**: `<any-value>` は任意のトークン列を受け入れるため、`d="garbage"` や `d="M 10 10 C 20 20"`（引数不足）を全て accept する。コメント「Validation not yet implemented」の通り、実質 placeholder 実装のまま。

### 修正方針

`<svg-path>` 型を真の SVG パス data parser に差し替える:

1. 空文字列は valid として受理（SVG2 仕様: 空は "no path"）
2. SVG2 BNF（https://svgwg.org/svg2-draft/paths.html#PathDataBNF）に基づいたパーサー:
   - コマンド文字認識（`M m L l H h V v C c S s Q q T t A a Z z`）
   - 各コマンドの引数 arity 検証
   - 数値形式の BNF 準拠

**Comment by YusukeHirao** [comment]
Closing as not-planned. The nu-validator benchmark shows this area as markuplint over-detection (32 of 34 fixtures are ml-only), not under-detection. Direction-corrected replacement: #3801.

**Comment by YusukeHirao** [comment]
Postscript: the intended replacement #3801 was also closed as not planned after the audit.

The benchmark's `ml-only` count on these SVG fixtures reflects nu-validator's leniency toward SVG 1.1 remnants (`version`, `baseProfile`, `xmlns:xlink`, `<font-face>` inside `<defs>`, `font-family="SVGFreeSansASCII,..."`), not a `<svg-path>` parsing issue — `<svg-path>` currently accepts any value via `<any-value>` and contributes zero violations on these fixtures. markuplint is spec-correct on every driver. There is nothing to fix here and nothing to fix in the replacement.

Both issues pre-dated the benchmark audit workflow in `tests/external/CLAUDE.md` that now catches this class of misreading. Leaving this closed.
